### PR TITLE
Optimizations

### PR DESCRIPTION
**Pull request by bording**
This includes several rather small changes that are focused around making a few small tweaks to prevent unnecessary memory allocations.

The first commit is also a refactoring that made sense given how the `ChannelProvider` and `ConfirmsAwareChannel` classes are actually the ones that need the `RoutingTopology`, instead of passing it to the` MessageDispatcher` first.

If there are any changes here that you'd prefer to have split into a separate PR, let me know and I can do that.

### CODE REVIEW

File: src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs [hunk 1]
```diff
@@ -74,42 +74,37 @@ public void RawSendInCaseOfFailure(IModel channel, string address, byte[] body,
...
             CreateExchange(channel, mainQueue);
             channel.QueueBind(mainQueue, mainQueue, string.Empty);
-
         }
 
-        static string ExchangeName(Type type)
-        {
-            return type.Namespace + ":" + type.Name;
-        }
+        static string ExchangeName(Type type) => type.Namespace + ":" + type.Name;
```
Comment (WojcikMike) [inline src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs]: For the ease of reading I would go with string interpolation here, but I am fine with leaving it as is.

Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs]: The problem with string interpolation is that, before .net 4.6, there's no nice way to make it culture invariant. In .net 4.6, we can use [FormattableString.Invariant](https://msdn.microsoft.com/en-us/library/system.formattablestring.invariant(v=vs.110).aspx). Whilst we're on .net 4.5 I think we should avoid string interpolation unless we explicitly want to use the current culture.

A micro-optimisation here would be to use `string.Concat()`, since the current use of two `+` operators forces an extra heap allocation.

File: src/NServiceBus.RabbitMQ/Sending/MessageDispatcher.cs [hunk 2]
```diff
@@ -21,14 +20,17 @@ public Task Dispatch(TransportOperations operations, ContextBag context)
 
             try
             {
-                var tasks = new List<Task>(operations.UnicastTransportOperations.Count() + operations.MulticastTransportOperations.Count());
+                var unicastTransportOperations = (List<UnicastTransportOperation>)operations.UnicastTransportOperations;
+                var multicastTransportOperations = (List<MulticastTransportOperation>)operations.MulticastTransportOperations;
```
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Sending/MessageDispatcher.cs]: I don't think we should make this cast. Although we know that *currently* core is passing in a `List<T>`, that might change, and the compiler will never tell us we have a problem. I'd be happy with:

```C#
var unicastTransportOperations = operations.UnicastTransportOperations as List<UnicastTransportOperation> ??
  operations.UnicastTransportOperations.ToList();
```

File: src/NServiceBus.RabbitMQ/Administration/QueueCreator.cs [hunk 3]
```diff
@@ -6,13 +6,13 @@
     class QueueCreator : ICreateQueues
     {
         readonly ConnectionFactory connectionFactory;
-        readonly IRoutingTopology topology;
+        readonly IRoutingTopology routingTopology;
```
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Administration/QueueCreator.cs]: Minor point, but I would have preferred to have this rename (and the similar one further down) as a separate commit to reduce the noise around the actual change. No need to amend though.

File: src/NServiceBus.RabbitMQ/Receiving/MessageConverter.cs [hunk 4]
```diff
@@ -128,7 +128,7 @@ static string ValueToString(object value)
             if (list1 != null)
             {
                 var list = list1;
-                return String.Join(";", list.Cast<object>().Select(ValueToString));
+                return String.Join(";", list.Cast<object>().Select(o => ValueToString(o)));
```
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Receiving/MessageConverter.cs]: What is the advantage of using the lambda here?

File: src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs (hunk 1 above)
Comment (SimonCropp) [inline src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs]: @adamralph given the concatenation of name and namespace can u explain how interpolation would have culture issues?

File: src/NServiceBus.RabbitMQ/Receiving/MessagePumpConnectionFailedCircuitBreaker.cs [hunk 5]
```diff
@@ -50,7 +50,7 @@ void CircuitBreakerTriggered(object state)
             if (Interlocked.Read(ref failureCount) > 0)
             {
                 Logger.WarnFormat("The circuit breaker for {0} will now be triggered", name);
-                triggerAction(lastException);
+                criticalError.Raise($"{name} connection to the broker has failed.", lastException);
```
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Receiving/MessagePumpConnectionFailedCircuitBreaker.cs]: Do we have a policy on string interpolation? See https://github.com/Particular/NServiceBus.RabbitMQ/pull/165#discussion_r68730535

File: src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs (hunk 1 above)
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs]: @SimonCropp in this case there would be no issue so we could indeed use string interpolation.

The slightly unsettling aspect for me is that in C# 5 we'd never use current culture for things like this. In moving to C# 6, in .NET 4.6 we can simply swap all our current concatenations and invariant `string.Format`s for bling strings wrapped in `Invariant()`, but in .NET 4.5 we're forced to make a decision about current vs invariant culture in each case, since bling string = current culture.

It would be less risky to stay away from bling strings in .NET 4.5 but if we feel the benefit of the syntax outweighs the risk then I'm happy to go with that.

File: src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs [hunk 6]
```diff
@@ -174,74 +174,60 @@ void Connection_ConnectionShutdown(object sender, ShutdownEventArgs e)
...
                 }
             }
+            catch (AlreadyClosedException ex)
+            {
+                Logger.Warn($"Attempt to acknowledge/reject message {messageId} failed because the channel was closed. The message will be requeued.", ex);
+            }
             catch (Exception ex)
             {
-                Logger.Warn($"Error while attempting to process message {messageId}. The message will be rejected.", ex);
+                Logger.Warn($"Error while attempting to process message {messageId}. The message will be requeued.", ex);
```
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs]: "rejected" -> "requeued"

Is this a change in behaviour?

File: src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs [hunk 7]
```diff
@@ -174,74 +174,60 @@ void Connection_ConnectionShutdown(object sender, ShutdownEventArgs e)
...
 
-                await RejectMessage(channel, message.DeliveryTag, messageId).ConfigureAwait(false);
+                try
+                {
+                    await RejectMessage(channel, message.DeliveryTag).ConfigureAwait(false);
+                }
+                catch (AlreadyClosedException) { }
```
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs]: We previously didn't swallow this exception. Is this a change in behaviour?

File: src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs [hunk 8]
```diff
@@ -174,74 +174,60 @@ void Connection_ConnectionShutdown(object sender, ShutdownEventArgs e)
...
-                Logger.Warn($"Attempt to reject message {messageId} failed because the channel was closed. The message will be requeued.", ex);
-            }
+                var s = (MessageState)messageState;
+
+                s.Channel.BasicReject(s.DeliveryTag, true);
+
+            }, new MessageState { Channel = channel, DeliveryTag = deliveryTag }, CancellationToken.None, TaskCreationOptions.DenyChildAttach, exclusiveScheduler);
         }
 
         public void Dispose()
```
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs]: In addition to my inline comments, it's difficult to tell if there may be other changes in behaviour in this commit. I think it's probably best to spin off this one into it's own PR.

File: src/NServiceBus.RabbitMQ/Sending/BasicPropertiesExtensions.cs [hunk 9]
```diff
@@ -32,12 +32,15 @@ public static void Fill(this IBasicProperties properties, OutgoingMessage messag
...
             {
-                properties.Type = message.Headers[NServiceBus.Headers.EnclosedMessageTypes].Split(new[]
+                var index = messageTypesHeader.IndexOf(',');
+
+                if (index > 0)
                 {
-                    ','
-                }, StringSplitOptions.RemoveEmptyEntries).FirstOrDefault();
+                    properties.Type = messageTypesHeader.Substring(0, index);
+                }
```
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Sending/BasicPropertiesExtensions.cs]: This looks like a change in behaviour. Previously, if there were no comma in the header, we'd set `properties.Type` to be the entire header. Now, we're only setting it if there is at least one comma.

File: src/NServiceBus.RabbitMQ/Configuration/ConnectionConfiguration.cs [hunk 10]
```diff
@@ -31,7 +31,7 @@ class ConnectionConfiguration
 
         public string CertPassphrase { get; set; }
 
-        public IDictionary<string, object> ClientProperties { get; } = new Dictionary<string, object>();
+        public Dictionary<string, object> ClientProperties { get; } = new Dictionary<string, object>();
```
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Configuration/ConnectionConfiguration.cs]: What is the reason for this change?

We gain the specificity of the type, but we lose the ability to change the concrete type.

File: src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs [hunk 11]
```diff
@@ -104,7 +104,7 @@ void SetupTypeSubscriptions(IModel channel, Type type)
                 }
             }
 
-            foreach (var exchangeName in type.GetInterfaces().Select(ExchangeName))
+            foreach (var exchangeName in type.GetInterfaces().Select(t => ExchangeName(t)))
```
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs]: What is the benefit of using the lambda?

File: src/NServiceBus.RabbitMQ/Configuration/ConnectionConfiguration.cs (hunk 10 above)
Comment (SimonCropp) [inline src/NServiceBus.RabbitMQ/Configuration/ConnectionConfiguration.cs]: irrelevant for an internal type

File: src/NServiceBus.RabbitMQ/Routing/DefaultRoutingKeyConvention.cs [hunk 12]
```diff
@@ -68,9 +36,28 @@ static string GetRoutingKey(Type type, string key = "")
...
+            return result;
+        }
+
+        static bool IsClrType(byte[] a1)
+        {
+            IStructuralEquatable structuralEquatable = a1;
+            return structuralEquatable.Equals(MsPublicKeyToken, StructuralComparisons.StructuralEqualityComparer);
+        }
+
+        static bool IsNServiceBusMarkerInterface(Type type) => type == typeof(IMessage) || type == typeof(ICommand) || type == typeof(IEvent);
```
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Routing/DefaultRoutingKeyConvention.cs]: Nitpick: in comparison with the original method above, this spills off the side off the diff and introduces a scroll bar. I'm not sure we have any policy on line length but I'd prefer to drop the body of the expression down to the next line.

File: src/NServiceBus.RabbitMQ/Configuration/ConnectionConfiguration.cs (hunk 10 above)
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Configuration/ConnectionConfiguration.cs]: Thanks, I missed that the class was internal. (I'm used to always seeing explicit `internal`.)

Agreed, concrete type is fine.

File: src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs [hunk 13]
```diff
@@ -74,42 +74,37 @@ public void RawSendInCaseOfFailure(IModel channel, string address, byte[] body,
...
-                    baseType = typeToProcess.BaseType;
-                }
+                CreateExchange(channel, ExchangeName(baseType));
+                channel.ExchangeBind(ExchangeName(baseType), ExchangeName(typeToProcess), string.Empty);
+                typeToProcess = baseType;
+                baseType = typeToProcess.BaseType;
             }
 
-            foreach (var exchangeName in type.GetInterfaces().Select(ExchangeName))
+            foreach (var exchangeName in type.GetInterfaces().Select(t => ExchangeName(t)))
```
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs]: What's the benefit of using the lambda?

**Comment by adamralph** [comment]
@bording I've reviewed all the commits and left comments inline.

File: src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs (hunk 1 above)
Comment (bording) [inline src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs]: >A micro-optimisation here would be to use string.Concat(), since the current use of two + operators forces an extra heap allocation.

This already happens automatically: https://ericlippert.com/2013/06/17/string-concatenation-behind-the-scenes-part-one/

For example, here's the IL for `ExchangeName()`:

```
.method private hidebysig static string ExchangeName(class [mscorlib]System.Type 'type') cil managed
{
    .maxstack 8
    L_0000: ldarg.0 
    L_0001: callvirt instance string [mscorlib]System.Type::get_Namespace()
    L_0006: ldstr ":"
    L_000b: ldarg.0 
    L_000c: callvirt instance string [mscorlib]System.Reflection.MemberInfo::get_Name()
    L_0011: call string [mscorlib]System.String::Concat(string, string, string)
    L_0016: ret 
}
```

Comment (bording) [inline src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs]: > . In moving to C# 6, in .NET 4.6 we can simply swap all our current concatenations and invariant string.Formats for bling strings wrapped in Invariant()

Is this a more general statement? I don't recall seeing any `String.Format` with specified cultures anywhere in our codebases.

File: src/NServiceBus.RabbitMQ/Sending/MessageDispatcher.cs (hunk 2 above)
Comment (bording) [inline src/NServiceBus.RabbitMQ/Sending/MessageDispatcher.cs]: Yeah, I was thinking this one would be a bit controversial, for the exact reason you pointed out. I didn't just go for it since the goal here is removing allocations and calling `ToList` would allocate.

However, that allocation means core has changed, and an allocation vs. blowing up is definitely preferable. Gives us time to react to the core change without having to have them perfectly in sync.

Will make this change.

File: src/NServiceBus.RabbitMQ/Administration/QueueCreator.cs (hunk 3 above)
Comment (bording) [inline src/NServiceBus.RabbitMQ/Administration/QueueCreator.cs]: I agree, and I do typically do renames in a separate commit, but I missed this one somehow. Didn't seem worth trying to go back and split it out once I noticed it, though.

File: src/NServiceBus.RabbitMQ/Receiving/MessageConverter.cs (hunk 4 above)
Comment (bording) [inline src/NServiceBus.RabbitMQ/Receiving/MessageConverter.cs]: The advantage here is that since this is a lambda that doesn't close over anything, the compiler generates code that creates and caches the delegate once and reuses it for every call.

This optimization doesn't happen you pass in a method group, so every call allocates a new delegate. There is an open PR on the Roslyn repo to change this behavior.

File: src/NServiceBus.RabbitMQ/Receiving/MessagePumpConnectionFailedCircuitBreaker.cs (hunk 5 above)
Comment (bording) [inline src/NServiceBus.RabbitMQ/Receiving/MessagePumpConnectionFailedCircuitBreaker.cs]: If we do, I'm not aware of one. From what I've seen, if anything it's "String Interpolation: GOOD".

File: src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs (hunk 7 above)
Comment (bording) [inline src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs]: No, because the `RejectMessage` method would either have not called `BasicReject` or would have caught the message.

Given the restructuring of the methods, this was needing to not throw an exception that would have previously been caught

File: src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs (hunk 6 above)
Comment (bording) [inline src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs]: Not really, given the restructuring of the methods and catch blocks. Previously if you got to that catch, something really weird would have happened, so there's a good chance your connection is already lost (which automatically requeues the message), The `RejectMessage` call used to have separate logging for that case, but it doesn't now. I changed the message to indicate that the message would be requeued because A) we call `BasicReject` and it succeeds, or B) the connection is already closed and requeued anyway.

File: src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs (hunk 8 above)
Comment (bording) [inline src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs]: :+1:  It would be pretty easy to split this off into it's own PR, and I can make sure its split into a couple commits to make it clearer.

File: src/NServiceBus.RabbitMQ/Sending/BasicPropertiesExtensions.cs (hunk 9 above)
Comment (bording) [inline src/NServiceBus.RabbitMQ/Sending/BasicPropertiesExtensions.cs]: Strictly speaking, yes this is a change, but it's one that I can't see impacting us, since the EnclosedMessageTypes header contains the AssemblyQualifiedName of the type, which is going to have a comma.

I could tweak this a bit to something like:

```csharp
if (index > -1)
{
    properties.Type = messageTypesHeader.Substring(0, index);
}
else
{
     properties.Type = messageTypesHeader
}
```

which would match match for the case you mention. If the string had a leading comma, the behavior would still be different. 

However, given that we're working with AssemblyQualifiedNames, does that matter?

File: src/NServiceBus.RabbitMQ/Configuration/ConnectionConfiguration.cs (hunk 10 above)
Comment (bording) [inline src/NServiceBus.RabbitMQ/Configuration/ConnectionConfiguration.cs]: The reason for the change is that enumerating a `Dictionary` is allocation-free, but allocating `IDictionary` isn't.

File: src/NServiceBus.RabbitMQ/Routing/DefaultRoutingKeyConvention.cs (hunk 12 above)
Comment (bording) [inline src/NServiceBus.RabbitMQ/Routing/DefaultRoutingKeyConvention.cs]: With indentation, the line is question is 143 characters long, and doesn't even reach half way on my monitor in Visual Studio. :smile: 

I've never been a huge fan of wrapping stuff to multiple lines just because, especially if it's a single expression.

If we were going to enforce some sort of line length I'd personally prefer it to be in relation to how the code looks in VS than how it shows up on GitHub.

**Comment by bording** [comment]
Ok, I've pushed up some new changes, primarily around reacting to the feedback.
 
I've also removed the message pump commit and am going to push that up as a separate PR.

@adamralph Even though some of my responses to your feedback are referencing "outdated commits", please let know if is anything else you have questions about.

I'm not entirely sure what, if anything, it will make sense to squash here, but I have included a couple fixups that would definitely be worth squashing.

File: src/NServiceBus.RabbitMQ/Receiving/MessageConverter.cs [hunk 14]
```diff
@@ -83,10 +83,9 @@ string DefaultMessageIdStrategy(BasicDeliverEventArgs message)
             return properties.MessageId;
         }
 
-        static Dictionary<string, string> DeserializeHeaders(BasicDeliverEventArgs message)
+        static Dictionary<string, string> DeserializeHeaders(IDictionary<string, object> headers)
```
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Receiving/MessageConverter.cs]: A good change for sure, but out of curiosity, how does it "Remove LINQ usage"?

File: src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs (hunk 1 above)
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs]: > This already happens automatically

Interesting, I didn't know that! Up to four operands, apparently.

> Is this a more general statement?

Yep, it's a general statement, very much so. Current culture is often (unknowingly) misused and in most cases, it just happens to work because there's nothing that in the formatting that would be affected. In most other projects I work on, we allow code analysis to tell us when we forget to specify the culture.

File: src/NServiceBus.RabbitMQ/Receiving/MessageConverter.cs (hunk 4 above)
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Receiving/MessageConverter.cs]: I learnt something again! Thanks.

File: src/NServiceBus.RabbitMQ/Sending/BasicPropertiesExtensions.cs (hunk 9 above)
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Sending/BasicPropertiesExtensions.cs]: What if someone is doing native integration by sending to RabbitMQ from a non-NSB endpoint? Could this header potentially be missing a comma in that case?

File: src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs (hunk 11 above)
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs]: Same as previous question, and answer given.

File: src/NServiceBus.RabbitMQ/Routing/DefaultRoutingKeyConvention.cs (hunk 12 above)
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Routing/DefaultRoutingKeyConvention.cs]: The problem with that is that everyone's VS is different, but everyone's GitHub is the same.

Anyway, we can continue this (hugely compelling) discussion elsewhere. :smile:

File: src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs (hunk 13 above)
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Routing/ConventionalRoutingTopology.cs]: Answered above.

File: src/NServiceBus.RabbitMQ/Receiving/MessageConverter.cs (hunk 14 above)
Comment (bording) [inline src/NServiceBus.RabbitMQ/Receiving/MessageConverter.cs]: `ToDictionary` is a LINQ extension method. I suppose a more precise commit message would have been "Remove ToDictionary usage from DeserializeHeaders"

Comment (bording) [inline src/NServiceBus.RabbitMQ/Receiving/MessageConverter.cs]: Also, this specific commit is a an additional fixup to be squashed into 0aacc7240952641d5996ac7b4bc90384392b2e7c I realized this morning that I could improve it a bit further by changing the parameter type, hence the fixup commit.

Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Receiving/MessageConverter.cs]: I can't see any use of `ToDictionary()` here.

Anyhow, once this is squashed into it's a moot point.

**Comment by adamralph** [comment]
@bording the changes look good to go!

Regarding squashing, I really like the original list of commits so I wouldn't want to squash any more than those. If it's not too much effort, I'd love for the new fix up commits to be squashed into the commit they are fixing up.

File: src/NServiceBus.RabbitMQ/Sending/BasicPropertiesExtensions.cs (hunk 9 above)
Comment (bording) [inline src/NServiceBus.RabbitMQ/Sending/BasicPropertiesExtensions.cs]: `BasicPropertiesExtensions.Fill` gets called when NSB is sending a message, so native integration isn't relevant here. For this to not have a comma, core would have to fundamentally change the value being stored in this header.

**Comment by bording** [comment]
Ok, all of the fixups have been squashed. I believe this covers all of the comment raised so far.

File: src/NServiceBus.RabbitMQ/Sending/BasicPropertiesExtensions.cs (hunk 9 above)
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Sending/BasicPropertiesExtensions.cs]: Ah, OK, makes sense then.

**Comment by adamralph** [comment]
:+1: awesome work @bording!
